### PR TITLE
Re-tools harbinger wraith

### PR DESCRIPTION
**Pull request by Bartimeus973**
## About the PR 
Harbinger portals now start off extremely weak. Using the ability again opens up a TGUI window allowing you to upgrade it.
portalhud
Upgrades can only be purchased when the harbinger is within 8 tiles of the portal. Purchasing an upgrade increases the price of subsequent upgrades and only one upgrade can be purchased every 15 seconds. If the portal is broken, all upgrades are lost , costs are reset and summoned creatures disappear.

Purchasing 3 upgrades or more allows the wraith to use a new "summon spectral guard" ability which can only be used in the portal's area of influence. Spawns a guard which slams in a 2 tile line whenever a human or cyborg comes close to it. 25 Health.
guardhit

Purchasing 8 upgrades or more blows up the lights in the portal's vicinity and slowly makes it spread void tiles around it.

Purchasing 20 upgrades gives the portal a new aura which inflicts any human who stays near it with a status which lowers maximum health and stamina until they get away from the portal.

## Why's this needed? 
Harbinger wraith lacks things to do compared to the other two wraith paths. The portal was overall fairly uninteresting and could just be left alone without much incentive to defend it. Giving it the option to be much stronger and customizable should make it more fun, useful and provides an incentive to defend it.
Ethereal guards provides the harbinger wraith with a way to directly defend their own portal, which they lacked until now. Also allows harbinger to solidify it's identity as a minion summoning, area denying wraith path.

```changelog
(u)Bartimeus973
(*)Harbinger wraith's 'summon portal' ability has been reworked and is now customizable! Also gives them a new ability to summon a ghostly guard to defend an area.
```

### CODE REVIEW

**Comment by Bartimeus973** [comment]
My limited knowledge of tgui means im not entirely sure how to resolve the conflict :/

**Comment by Bartimeus973** [comment]
Turns out vs code was setting double quotes to single quotes for no good reason. Fixed it.

File: code/datums/abilities/wraith.dm [hunk 1]
```diff
@@ -1597,84 +1597,90 @@ ABSTRACT_TYPE(/datum/targetable/wraithAbility/curse)
...
+	cooldown = 120 SECONDS
 
 	cast()
 		if (..())
-			return 1
+			return TRUE
 
 		var/turf/T = get_turf(holder.owner)
 		if (isturf(T) && istype(T,/turf/simulated/floor))
+			for (var/turf/simulated/wall/W in T)
```
Comment (pali6) [inline code/datums/abilities/wraith.dm:1611]: turfs can't be in a turf, this check does nothing

File: code/datums/abilities/wraith.dm [hunk 2]
```diff
@@ -1597,84 +1597,90 @@ ABSTRACT_TYPE(/datum/targetable/wraithAbility/curse)
...
+				boutput(holder.owner, "<span class='notice'>You cannot place your portal in a wall!</span>")
+				return TRUE
+			for (var/obj/window/W in T)
```
Comment (pali6) [inline code/datums/abilities/wraith.dm:1614]: Are you sure just windows? Don't we want to check for like all dense objects or something?

File: code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+
+	disposing()
+		if (src.master != null)
+			src.master.guard_amount --
+		. = ..()
+
+	attackby(obj/item/P as obj, mob/living/user as mob)
+		src._health -= P.force
+		attack_particle(user,src)
```
Comment (pali6) [inline code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm:49]: Isn't there some proc for modifying _health? If not, there probably should be.

File: code/modules/status_system/statusEffects.dm [hunk 4]
```diff
@@ -2200,3 +2200,34 @@
...
+			owner.delStatus("portal_weakness")
+		else
+			H = owner
+			H.max_health -= max_health
+			health_update_queue |= H
+			H.add_stam_mod_max("portal_weakness", max_stam)
+
+	onRemove()
+		. = ..()
+		if (ishuman(owner))
```
Comment (pali6) [inline code/modules/status_system/statusEffects.dm:2227]: Should probably `H = null` after this. Or just don't have H as a member variable but as a local on both procs.

File: code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	proc/attack(var/mob/living/M)
+		var/list/attacked = list()
+		var/direction = get_dir(src, M)
+		src.set_dir(direction)
+		var/turf/one = get_step(src, direction)
+		new/obj/decal/guard_attack_marker(one, src.attack_delay)
+		var/turf/two = get_step(one, direction)
+		new/obj/decal/guard_attack_marker(two, src.attack_delay)
+		SPAWN(src.attack_delay)
+			if (src)
```
Comment (pali6) [inline code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm:76]:
```suggestion
			if (!QDELETED(src))
```
src can't disappear like this, but the object can become qdeleted which you should check

File: code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+				for(var/turf/T in list(one, two))
+					for(var/atom/A in T)
+						if(A in attacked) continue
+						if(ismob(A) && !isobserver(A))
+							var/mob/target_mob = A
+							target_mob.TakeDamageAccountArmor("All", 18, 0, 0, DAMAGE_BLUNT)
+							attacked += A
+							playsound(src.loc, 'sound/impact_sounds/Generic_Hit_2.ogg', 80, 1)
+							src.visible_message("<span class='combat'>[src] slams the ground and hits [target_mob]!</b></span>")
+			return
```
Comment (pali6) [inline code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm:86]:
```suggestion
```

File: code/datums/abilities/wraith.dm [hunk 7]
```diff
@@ -1597,84 +1597,90 @@ ABSTRACT_TYPE(/datum/targetable/wraithAbility/curse)
...
+			return TRUE
+		if (the_wraith.guard_amount >= the_vortex.mob_value_cap)
+			boutput(holder.owner, "<span class='notice'>You reached maximum guard capacity! Upgrade your portal's summon capacity to summon more!</span>")
+			return TRUE
+		var/turf/T = null
+		T = get_turf(target)
+		if (!T || !istype(T,/turf/simulated/floor))
+			boutput(holder.owner, "<span class='notice'>You cannot use this here!</span>")
+			return TRUE
+		for (var/turf/simulated/wall/W in T)
```
Comment (pali6) [inline code/datums/abilities/wraith.dm:1669]: Some as first comment.

File: code/datums/abilities/wraith.dm [hunk 8]
```diff
@@ -1597,84 +1597,90 @@ ABSTRACT_TYPE(/datum/targetable/wraithAbility/curse)
...
+			boutput(holder.owner, "<span class='notice'>You cannot place a guard in a wall!</span>")
+			return TRUE
+		for (var/obj/window/W in T)
```
Comment (pali6) [inline code/datums/abilities/wraith.dm:1672]: Same as my second comment.

**Review by pali6 — CHANGES_REQUESTED** [review]

File: code/datums/abilities/wraith.dm (hunk 2 above)
Comment (Bartimeus973) [reply code/datums/abilities/wraith.dm]: I'll now check for objects with density = 1

File: code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm (hunk 3 above)
Comment (Bartimeus973) [reply code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm:49]: It seems there isnt, and that each machinery has their own way of treating damage. Would you prefer that i bundled a proc for machineries to treat damage in this pr, or should it be atomized?

Comment (Bartimeus973) [reply code/modules/antagonists/wraith/objs/machinery/harbinger_guard.dm:48]: Actually, changeHealth() does exactly that, and also triggers destruction when health reaches zero. Implemented it and added it to the portal too.

File: code/modules/antagonists/wraith/objs/machinery/vortex_wraith.dm [hunk 9]
```diff
@@ -141,36 +138,223 @@
...
-			if (/obj/critter/gunbot/heavy)
-				return 15
-			if (/obj/critter/ancient_thing)
-				return 7
-			if (/obj/critter/mechmonstrositycrawler)
-				return 4
-			else	//You never know, lets give an average point cost
-				return 6
+	attack_hand(mob/user)
+		return
```
Comment (frawhst) [inline code/modules/antagonists/wraith/objs/machinery/vortex_wraith.dm:148]: Why is this here

File: code/modules/antagonists/wraith/objs/machinery/vortex_wraith.dm [hunk 10]
```diff
@@ -141,36 +138,223 @@
...
+					. = TRUE
+					return
+				else
+					if (src.spawn_rate <= 25 SECONDS)
+						boutput(usr, "<span class='notice'>Maximum spawn rate already achieved!</span>")
+						. = TRUE
+						return
+					else
+						src.spawn_rate -= 3 SECONDS
+						src.last_upgrade = TIME + src.upgrade_cooldown
```
Comment (frawhst) [inline code/modules/antagonists/wraith/objs/machinery/vortex_wraith.dm:213]: relevant line to above comment (no changes needed here other than the var rename)
